browser/src/layer/marker/TextInput.js

```diff
@@ -680,7 +681,20 @@ L.TextInput = L.Layer.extend({
 		this._lastContent = content;
 
 		if (newText.length > 0) {
+			if (this._map.getDocType() === 'spreadsheet' && content.length === 1 && ev.inputType === 'insertText') {
+				newText.push('37'); //% sign
+				content = newText;
+				this._lastContent = newText;
+				var percentText = this._textArea.value;
+				this._textArea.value = percentText.slice(0,2) + '%' + percentText.slice(2);
+				this._hasAutoPercent = true;
+			}
+
 			this._sendText(this.codePointsToString(newText));
+
+			if (content.length === 2 && this._hasAutoPercent) {
+				this._setCursorPosition(2);
```

Comment:
IMO, this only has to be done in the same event when the percent was appended.

Imagine this:
1. I type "1"
2. It appends "%", sets _hasAutoPercent, and moves the cursor to position 2 (OK)
3. I type "2" (the current text is 12%)
4. I press "Del"

At this point, the text is "12", and its length is 2, and _hasAutoPercent is true. It will put the cursor between 1 and 2, while it should stay after the 2.

So you don't need _hasAutoPercent.

Reply:
Agree, not using any "state" variable would be perfect :)